airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_source.py

```diff
@@ -0,0 +1,165 @@
+from airbyte_cdk.sources.concurrent_source.partition_generation_completed_sentinel import PartitionGenerationCompletedSentinel
+from airbyte_cdk.sources.concurrent_source.thread_pool_manager import ThreadPoolManager
+from airbyte_cdk.sources.message import InMemoryMessageRepository, MessageRepository
+from airbyte_cdk.sources.streams.concurrent.abstract_stream import AbstractStream
+from airbyte_cdk.sources.streams.concurrent.partition_enqueuer import PartitionEnqueuer
+from airbyte_cdk.sources.streams.concurrent.partition_reader import PartitionReader
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+from airbyte_cdk.sources.streams.concurrent.partitions.types import PartitionCompleteSentinel, QueueItem
+from airbyte_cdk.sources.utils.slice_logger import DebugSliceLogger, SliceLogger
+
+
+class ConcurrentSource:
+    """
+    A Source that reads data from multiple AbstractStreams concurrently.
+    It does so by submitting partition generation, and partition read tasks to a thread pool.
+    The tasks asynchronously add their output to a shared queue.
+    The read is done when all partitions for all streams were generated and read.
+    """
+
+    DEFAULT_TIMEOUT_SECONDS = 900
+
+    @staticmethod
+    def create(
+        num_workers: int,
+        initial_number_of_partitions_to_generate: int,
+        logger: logging.Logger,
+        slice_logger: SliceLogger,
+        message_repository: MessageRepository,
+        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
+    ) -> "ConcurrentSource":
+        threadpool = ThreadPoolManager(
+            concurrent.futures.ThreadPoolExecutor(max_workers=num_workers, thread_name_prefix="workerpool"), logger, num_workers
+        )
+        return ConcurrentSource(
+            threadpool, logger, slice_logger, message_repository, initial_number_of_partitions_to_generate, timeout_seconds
+        )
+
+    def __init__(
+        self,
+        threadpool: ThreadPoolManager,
+        logger: logging.Logger,
+        slice_logger: SliceLogger = DebugSliceLogger(),
+        message_repository: MessageRepository = InMemoryMessageRepository(),
+        initial_number_partitions_to_generate: int = 1,
+        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
+        raise_exception_on_missing_stream: bool = True,
+        **kwargs: Any,
+    ) -> None:
+        """
+        :param threadpool: The threadpool to submit tasks to
+        :param message_repository: The repository to emit messages to
+        :param max_number_of_partition_generator_in_progress: The initial number of concurrent partition generation tasks. Limiting this number ensures will limit the latency of the first records emitted. While the latency is not critical, emitting the records early allows the platform and the destination to process them as early as possible.
```

Comment:
Nit: some of the params are missing and/or have been renamed since.

Regarding the comment for `max_number_of_partition_generator_in_progress` / `initial_number_partitions_to_generate` - looks like the idea is that we'll generate one partition per stream, and so this variable basically is there to say "you can start processing records after `<initial_number_partitions_to_generate>` partitions have been created"; if there are tons of streams, it means that we're allowed to start processing records before all of the stream partitions have been generated. Is that right?

It's not obvious to me that we'd ever want this to be more than one - did you have a use case in mind or is this here for configurability just in case?

Reply:
> this variable basically is there to say "you can start processing records after <initial_number_partitions_to_generate> partitions have been created"

that's not quite what it says. records can be started after N tasks to generate partitions were submitted to the threadpool. 

If the N == one and generating the partitions takes a long time (eg if we need to read all records from a parent stream), then the workers might be idle for some time.

If the N > number of workers it might take some time for the first records to be polled from the queue since the first many items will be partitions to process. I mostly want to avoid partition generation tasks from using all the workers and preventing them from actually processing the partitions.

An alternative implementation might be to submit the first partition generation before polling from the queue, and submit the others when we read the first record.

Reply:
Okay thanks for the explanation. Tbh I'm not entirely sold on the extra complexity that this requires. If we we were using coroutines/asynico here we wouldn't have this issue because the executor would be able to flit between tasks while I/O is happening, whereas the standard threads are blocking. Even though your code helps alleviate some of this pressure we're always going to have the problem of threads idling during I/O as long as we're using standard threads. We could revisit using asyncio, though I recognize it might require significant changes. Alternatively, I think we could get a similar efficiency boost by replacing standard threads with green threads, as they offer context switching during I/O operations with less overhead. IMO that's preferable to adding complexity that doesn't entirely solve the problem.

Reply:
We synced offline.

The crux of the issue comes from our threadpool based approach. If using N workers, and they are all processing long partition generation tasks, no workers are available for emitting records from those partitions, which affects the overall sync performance.

@clnoll's suggestion of using asyncio is great as it should allow us to simplify the flow and might also improve the performance. We'll prioritize a spike to evaluate the effort and go from there.